### ISSUE TITLE
Kodi switches to incorrect whitelisted resolution

### ISSUE THREAD

**Issue by vboginskey** [xbmc/xbmc#19037]
### Describe the bug
Here is a clear and concise description of what the problem is:

When playing a 3840x2160 video on a 3840x2160 display, Kodi switches to 4096x2160.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

Kodi should switch to 3840x2160.

## Actual Behavior

Kodi switches to 4096x2160.

## Possible Fix

https://github.com/xbmc/xbmc/blob/66d5ea75bd3d67a320fcbdca81261381951a7cc8/xbmc/windowing/Resolution.cpp#L134

The above comparison allows 4096x2160 to "exactly" match 3840x2160 because, at evaluation time,
```
height = 2160
width = 3840
info.iScreenHeight = 2160
info.iScreenWidth = 4096
```

The condition will therefore evaluate to true.

### To Reproduce

1. Whitelist all 4096x2160 and 3840x2160 modes
2. Play 3840x2160 video
3. Press `z` and observe the incorrect resolution

### Debuglog

The debuglog can be found here: http://paste.kodi.tv/acizezajuk.kodi

### Screenshots 
Here are some links or screenshots to help explain the problem:

Debug snip:
```
2021-01-07 22:56:04.643 T:4948 INFO <general>: [WHITELIST] Searching the whitelist for: width: 3840, height: 2160, fps: 24.000, 3D: false
2021-01-07 22:56:04.643 T:4948 DEBUG <general>: [WHITELIST] Searching for an exact resolution with an exact refresh rate
2021-01-07 22:56:04.643 T:4948 DEBUG <general>: [WHITELIST] Matched an exact resolution with an exact refresh rate Generic PnP Monitor #1: 4096x2160 @ 24.00Hz (328)
2021-01-07 22:56:04.643 T:4948 INFO <general>: Display resolution ADJUST : Generic PnP Monitor #1: 4096x2160 @ 24.00Hz (328) (weight: 0.000)
```

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

Forum post: https://forum.kodi.tv/showthread.php?tid=359682

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] OSX
 - [x] Windows
 - [ ] Windows UWP

 - Operating system version/name: Windows 10 Pro, 2020H2 (19042.685)
 - Kodi version: 19.0 - Beta2 (18.9.821)

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by fritsch** [comment]
Could you name the reason why you have 4096x2160 in your whitelist?

**Comment by vboginskey** [comment]
I whitelisted all the modes supported by my hardware.

**Comment by lrusak** [comment]
Only whitelist the modes you want to use.

**Comment by vboginskey** [comment]
I want to use 4096x2160 for 4096x2160 videos. I would expect Kodi to match the resolution to the content exactly, not choose the first one that matches its height.

**Comment by lrusak** [comment]
> I want to use 4096x2160 for 4096x2160 videos. I would expect Kodi to match the resolution to the content exactly, not choose the first one that matches its height.

I believe this is the intended behavior. See https://github.com/xbmc/xbmc/pull/14132

Just disable 4096x2160 and let the image be scaled to 3840x2160.

I know this probably seems odd but there is really no 100% solution for all cases.

**Comment by vboginskey** [comment]
Thanks for the context. To be clear, I know I can fix the problem by disabling 4096x2160. But, I don't want to just fix my issue and move on, I want to help make this software better for other people.

Reading through the examples in the linked PR - they're talking about an error of a few dozen pixels at most, and only in cases where there is no existing match for the video resolution. What we are dealing with here is Kodi choosing squarely the wrong resolution when an exact match is available. Technical users like ourselves can identify the error and "manually select desired resolution", but how many people are out there unknowingly watching 16:9 content on their 16:9 OLEDs with stretched pixels and vertical bars as a result of this behavior?

To put it another way, would you agree that is logic is strictly better than what we have today:
1. If: an exact resolution and refresh rate are available, choose that
2. Else: fall back to old behavior

I can scrape together a PR that accomplishes this, but I've never contributed to this repo before and I'd rather not waste my time if that contribution is not welcome.

**Comment by lrusak** [comment]
The whitelist really is for users that need to fine tune their resolution switching if the default logic doesn't work for them.

The whitelist was also designed to use lower resolutions to avoid upscaling in shaders and allowing the monitors native scalers to do the work.

Again there really isn't a 100% solid solution here and by changing this behavior likely something else will break.

Kodi will scale the media to whatever resolution is currently being used if the media doesn't match the resolution anyways. So if users want to watch all their media at 1080p60 they are free to do so.

So you are free to make a PR but I cannot say if it will be accepted or not without seeing the solution.

**Comment by vboginskey** [comment]
> The whitelist really is for users that need to fine tune their resolution switching if the default logic doesn't work for them.

The default logic fixes the refresh rate which leads to visible stutter in most content. It's why I started using the whitelist.

> Again there really isn't a 100% solid solution here and by changing this behavior likely something else will break.

I'm not advocating for a 100% solution, I'm advocating for choosing the exact resolution from the whitelist if one is available. I can't imagine why any user would want Kodi to choose the wrong whitelisted resolution over the exact match whitelisted resolution.

I can't help but feel like the response is just defending the existing logic and not considering my argument on its merits.

**Comment by DaVukovic** [comment]
Simply create that PR and let's discuss possible solutions on existing (proposed) code.

**Comment by salim-b** [comment]
I support what @vboginskey described.

My particular case:

1. I have a display (actually a Sony VPL-VW570ES) with a native resolution of 4096x2160. It displays 3840x2160 input signals pixel-perfect, too (by leaving a width of 256 pixels blank, 128 on each side). I want the display instead of Kodi to perform the upscaling whenever possible, so I whitelisted all available resolutions.

2. I get some weird resolution choices by Kodi:

    1. **For media in 3840x2160 Kodi chooses an output resolution of 4096x2160.** -> Doesn't make any sense to me since the media's resolution would either need to be cropped in height or otherwise need to be stretched in width (changing the aspect ratio) in order to fit in the output resolution of 4096x2160.
    
    2. **For media in 3840x1604 Kodi chooses an output resolution of 3840x2160.** -> Ok, but suboptimal depending on your preference since the media could be scaled proportionally to the max output resolution of 4096x2160 (i.e. without affecting the media's aspect ratio).
    
After studying `xbmc/xbmc/windowing/Resolution.cpp` I think the logic in there is somewhat skewed.

The `<=`'s in the following if-condition does simply not differentiate between an _exact_ resolution match of media and output (2D resolution congruence) and a media resolution that just "fits in" the output resolution by only exactly matching width _or_ height (1D resolution congruence):

https://github.com/xbmc/xbmc/blob/66d5ea75bd3d67a320fcbdca81261381951a7cc8/xbmc/windowing/Resolution.cpp#L134-L135

Therefore I've submitted https://github.com/xbmc/xbmc/pull/19119 fixing the first case of the two I described above.

The second case could also be tackled of course but I believe this would only make sense in companion with giving the user more control (i.e. introducing another display option in the settings) because some people might prefer seeing some unused space on their displays over Kodi scaling the input media.

**Comment by fritsch** [comment]
Fix it. Send a Pull Request. Writing walls of text and complaining about flawed logic, has now promoted you to make it better.

So, the world was waiting for you to fix this flawed logic. The stage is yours.

**Comment by fritsch** [comment]
I comments in your PR: https://github.com/xbmc/xbmc/pull/19119/commits/480fc8608417d9d2193cb6ceab8db0d87ed9d7cb#diff-e899258d7761060e9bb26f447566bb60a5f8d7c138565c5260f8bb4ac1b062a1R151 

you broke 1920x800 for example and others ... your logic seems flawed.

**Comment by salim-b** [comment]
@fritsch Thanks for your review. I've updated the PR, please have a look.

> And then there are users whitelisting 4096 with their TV while even knowing that this is not the native resolution and are so stubborn to just program it in, while kodi with it's non linear scaling would nicely fit it into 3840.

I'm not really sure if I correctly understand what you're referring to, but I think the code shouldn't contain quirks for obvious configuration mistakes...?

**Comment by fritsch** [comment]
My point (as also descriped in the PR) says: Don't try to quirk without considering all cases. The Whitelist was done to give users a chance to influence automatisms. Adding a 4096x2160 resolution to a whitelist while the native resolution of the TV is 3840x2160 makes absoluteley zero sense and what you wanted to achieve, e.g. pixel perfect output will be nuts.

If you want to solve it more scientifically, keep in mind: you won't find a solution for every case, there will still be someone complaining.

Quite good examples are "old" 4:3 or similar modes, like 1600x1200 or 1024x768 and 1280x768 - but let's leave them shortly out.

What we could agree on I think is the following:
- If the image matches the width of a resolution while the height of the image still fits into this resolution, it is a nice match. Example: Resolution 1920x1080: 1920x800 and 1920x816 videos should be display there.
- If video file perfectly "matches" into available resolutions, we can save us this discussion. Problems start when things don't directly match, e.g. macroblock widths and width / heights get mixeed. E.g. a silly demuxer tells a h264 video is 1088 in height while it only is 1080. This is where this +8 workaround is coming from ... in the first place.

So: What could one do to get it "better":
- Allow user to add / remove resolutions from the list like it is now
- Change the matching algorithm to be simpler but also more generic. Having 8 loops won't help anywhere.
- Prefer width over height for landscape videos.

So, if you really want to change something, try euclidian distance with width weighted higher (lower value at the end therefore worse penalty term, see * 2 below) than height and choose the minimum.

E.g. 1920x800 input, choose 1920x1080, because: (2 * (1920-1920))^2 + (1080-800)^2 is a better match than (2 * (3840-2160))^2 + (2160-800)^2). 

And even this will get complicated as low resolution width are quite near to each other and will fool your results. Famous example: 2560*1600 - would it choose 3840x2160 oder 1920x1080? :-)

I hope you now understand why tuning has limits.

For above bug report: Plain remove your 4096 resolution from the whitelist. It's idiotic and exactly those special cases will make this code very hard to write and to debug.

**Comment by salim-b** [comment]
Ok, thanks for the detailed explanation, makes sense to me except for one fundamental assumption of yours being wrong in certain cases: Some devices (as the Sony VPL-VW570ES projector in my case) act as they had multiple native resolutions. While the projector panel's full resolution is actually 4096x2160, it also displays 3840x2160 input signals completely pixel-perfect (by leaving blank spaces on the left and right as I described above).

Besides that I'd totally vote for

> Change the matching algorithm to be simpler but also more generic. Having 8 loops won't help anywhere.

But I guess I'm not the right person to do this since I'm neither familiar with the codebase nor proficient in C++...

**Comment by fritsch** [comment]
How do the blank spaces work?

You know, when kodi outputs 3840x2160 on a 4096 x 2160 screen content, it will actually already touch each and every pixel so that you have 4096xsomething output. Picture is broken the very moment you output something to it, pixel data has been altered.

**Comment by salim-b** [comment]
> How do the blank spaces work?

Kinda like letterboxing works, but horizontally. In case of a 3840x2160 input signal, the projector just outputs an image with the same height as with an 4096x2160 input signal _but_ with 256 pixels less width. No scaling, no stretching. With the advantage that the blank space is not just black picture but really dark (i.e. no direct light from the lamp; IDK how exactly this is achieved, I guess some kind of aperture is used).

> You know, when kodi outputs 3840x2160 on a 4096 x 2160 screen content, it will actually already touch each and every pixel so that you have 4096xsomething output.

🤨 I can't quite follow I'm afraid. I think what you're describing is my original concern and the reason why I stumbled upon this issue: I do _not_ want Kodi to output a 4096x2160 signal in case of 3840x2160 content because that would necessarily imply the content either being

1. cropped (bad),

2. changed in aspect ratio (bad) or

3. rendered with vertical blank (black) bars. This seems the best strategy but is a) not done right now AFAICT from my experience and b) still not desirable in my case since these black bars would be illumined by the projector while the same picture isn't illumined with a 3840x2160 output signal).

**Comment by fritsch** [comment]
Cool, then we agree.

**Comment by fritsch** [comment]
That would work for you, right: https://paste.ubuntu.com/p/tgDZYTNZ5C/ ?

**Comment by salim-b** [comment]
> That would work for you, right: https://paste.ubuntu.com/p/tgDZYTNZ5C/ ?

See https://github.com/xbmc/xbmc/pull/19119#issuecomment-767100935 :slightly_smiling_face:
